src/basic/string-util.rs

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+use std::ptr;
+
+/// Extract the i'nth line from the specified string. Returns > 0 if there are more lines after
+/// that, and == 0 if we are looking at the last line or already beyond the last line. As special
+/// optimization, if the first line is requested and the string only consists of one line we return
+/// NULL, indicating the input string should be used as is, and avoid a memory allocation for a
+/// very common case.
+/// # Safety
+/// This function expects a non-NULL 'ret' parameter, and will dereference it without checking.
```

Comment:
This is Markdown; they usually leave blank lines around the headings, like this (from std):

```rust
/// Reads the value from `src` without moving it. This leaves the
/// memory in `src` unchanged.
///
/// Unlike [`read`], `read_unaligned` works with unaligned pointers.
///
/// # Safety
///
/// Behavior is undefined if any of the following conditions are violated:
///
/// * `src` must be [valid] for reads.
///
/// * `src` must point to a properly initialized value of type `T`.
```

Also `'ret'` should be <code>&#96;ret&#96;</code>, etc.

Reply:
thanks, fixed